misc/connection-limits/src/connection_limits/memory.rs

```diff
@@ -0,0 +1,181 @@
+                    let process_memory_bytes = self.mem_tracker.total_bytes();
+                    if process_memory_bytes > max_allowed_bytes {
+                        return Err(ConnectionDenied::new(MemoryUsageLimitExceeded {
+                            process_memory_bytes,
+                            max_allowed_bytes,
+                        }));
+                    }
+                }
+            }
+            EstablishedIncoming | EstablishedOutgoing | EstablishedPerPeer | EstablishedTotal => {}
+        };
+
+        Ok(())
+    }
+}
+
+/// A connection limit has been exceeded.
+#[derive(Debug, Clone, Copy)]
+pub struct MemoryUsageLimitExceeded {
+    pub process_memory_bytes: usize,
+    pub max_allowed_bytes: usize,
+}
+
+impl std::error::Error for MemoryUsageLimitExceeded {}
+
+impl fmt::Display for MemoryUsageLimitExceeded {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "pending connections are dropped, process memory usage limit exceeded: process memory: {} bytes, max allowed: {} bytes",
+            self.process_memory_bytes,
+            self.max_allowed_bytes,
+        )
+    }
+}
+
+#[derive(Debug)]
+struct ProcessMemoryUsageTracker(Arc<RwLock<ProcessMemoryUsageTrackerInner>>);
+
+impl ProcessMemoryUsageTracker {
+    fn new() -> Self {
+        Self(Arc::new(RwLock::new(ProcessMemoryUsageTrackerInner::new())))
+    }
+
+    fn refresh_if_needed(&self) {
+        let need_refresh = self.0.read().need_refresh();
+        if need_refresh {
+            self.0.write().refresh();
+        }
+    }
+
+    fn total_bytes(&self) -> usize {
+        let guard = self.0.read();
+        guard.physical_bytes + guard.virtual_bytes
+    }
+}
+
+#[derive(Debug)]
+struct ProcessMemoryUsageTrackerInner {
+    last_checked: Instant,
+    physical_bytes: usize,
+    virtual_bytes: usize,
+}
+
+impl ProcessMemoryUsageTrackerInner {
+    fn new() -> Self {
+        let stats = memory_stats::memory_stats();
+        if stats.is_none() {
+            log::warn!("Failed to retrive process memory stats");
+        }
+        Self {
+            last_checked: Instant::now(),
+            physical_bytes: stats.map(|s| s.physical_mem).unwrap_or_default(),
+            virtual_bytes: stats
+                .map(|s: memory_stats::MemoryStats| s.virtual_mem)
+                .unwrap_or_default(),
+        }
+    }
+
+    fn need_refresh(&self) -> bool {
+        const PROCESS_MEMORY_USAGE_CHECK_INTERVAL: Duration = Duration::from_millis(1000);
+
+        self.last_checked + PROCESS_MEMORY_USAGE_CHECK_INTERVAL < Instant::now()
+    }
+
+    fn refresh(&mut self) {
+        if let Some(stats) = memory_stats::memory_stats() {
+            self.physical_bytes = stats.physical_mem;
+            self.virtual_bytes = stats.virtual_mem;
+        } else {
+            log::warn!("Failed to retrive process memory stats");
+        }
+        self.last_checked = Instant::now();
+    }
```

Comment:
We should do this with a proper timer that can register a waker within `NetworkBehaviour::poll`.

Reply:
@thomaseizinger what would be the benefit? Say that the local node has no new connections for 5 minutes, neither inbound nor outbound. In such case there is no need to refresh the memory limit, no?

Reply:
> @thomaseizinger what would be the benefit? Say that the local node has no new connections for 5 minutes, neither inbound nor outbound. In such case there is no need to refresh the memory limit, no?

Yeah, if it is reasonably cheap, we should just get the latest stats upon every connection.

Reply:
@thomaseizinger @mxinden 
The [`memory_stats`](https://docs.rs/memory-stats/1.1.0/memory_stats/fn.memory_stats.html) function is not cheap, the [doc](https://docs.rs/memory-stats/latest/memory_stats/#caveats) says
> Getting accurate memory usage on Linux is fairly expensive and not always possible. This crate always attempts to use the statistics from [/proc/self/smaps](https://man7.org/linux/man-pages/man5/proc.5.html#:~:text=See%20user_namespaces%287%29.-,/proc/%5Bpid%5D/smaps,-%28since%20Linux%202.6.14) if avaliable. However, since support for /proc/self/smaps might not be compiled in on all kernels, this crate will also use the faster but less accurate statistics from [/proc/self/statm](https://man7.org/linux/man-pages/man5/proc.5.html#:~:text=by%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20waitpid%282%29.-,/proc/%5Bpid%5D/statm,-Provides%20information%20about) as a fallback.


> If speed is needed over accuracy, the always_use_statm feature can be enabled to always use the /proc/self/statm statistics.

I run cargo bench with below code
```rust
extern crate test;
    use test::Bencher;

    use super::*;

    #[bench]
    fn bench(bencher: &mut Bencher) {
        let  _memories = Vec::<u8>::with_capacity(128 * 1024 * 1024);
        bencher.iter(|| {
            let _ = memory_stats();
        });
    }
```

`cargo +nightly bench`
```console
test bench::bench ... bench:      53,453 ns/iter (+/- 6,062)
```

`cargo +nightly bench --features always_use_statm`
```console
test bench::bench ... bench:       1,725 ns/iter (+/- 222)
```

Please let me know your suggestion. I see 3 options,
1. Proactively(say, 1s) refresh mem stats without `always_use_statm` feature.
2. Passively refresh mem stats upon every connection with a expiration(say, 1s) (The current implementation in this PR) 
3. Get the latest stats upon every connection with `always_use_statm` feature

Reply:
I tried to do some research on how inaccurate `/proc/self/statm` is but couldn't find much. I'd say for our purpose it would be fine? Given that it makes the code much simpler, I'd lean towards always using the fast version and simply getting them on-demand.

>  If speed is needed over accuracy, the always_use_statm feature can be enabled to always use the /proc/self/statm statistics.

This is actually bad design because if we activate this feature, every other crate in our user's dependency tree will also use this code path. I opened an issue with them: https://github.com/Arc-blroth/memory-stats/issues/6

Reply:
@thomaseizinger That makes sense, will switch to 3 once the issue is resolved.

Reply:
After thinking about it, I'm not sure if ~2000ns per function call is trivial enough and I have not tested on all platforms (Windows, OSX).  Would it be safer to still use approach 2 with a smaller refresh interval (e.g. 100ms) in case a swarm is dealing with high connection per second?

Reply:
> After thinking about it, I'm not sure if ~2000ns per function call is trivial enough and I have not tested on all platforms (Windows, OSX). Would it be safer to still use approach 2 with a smaller refresh interval (e.g. 100ms) in case a swarm is dealing with a lot of connections?

Sure, let's keep the refresh approach then :)

Reply:
Fixed